### PR TITLE
Fix tooltip positioning with regards to viewports

### LINKED ISSUE #14195: Tooltip layout wrong when elements outside of viewport
look here:
http://jsbin.com/yalohofo/1/

The layout of the tooltips gets screwed up, if the tooltips are created when outside of the viewport !

### PR DESCRIPTION
**Pull request by cvrebert**
A rebased version of #14226.
Fixes #14195.

### CODE REVIEW

**Comment by ericmartel** [comment]
@cvrebert Excellent thank you!  I'll work on my git-fu for the next submits :)

File: js/tests/unit/tooltip.js [hunk 1]
```diff
@@ -680,6 +680,42 @@ $(function () {
     $styles.remove()
   })
 
+  test('should not be adjust position when scrolling', function () {
```
Comment (hnrch02) [inline js/tests/unit/tooltip.js]: Still includes the grammar mistake you pointed out before :smile:

Comment (ericmartel) [inline js/tests/unit/tooltip.js]: I'm surprised I didn't fix this one, my bad!  Now it seems in @cvrebert 's hands, sorry about that!

Comment (hnrch02) [inline js/tests/unit/tooltip.js]: Yeah, my comment was directed at @cvrebert, should've added his handle to the post.

Comment (cvrebert) [inline js/tests/unit/tooltip.js]: @ericmartel So, would I be correct in summarizing this test as "scrolling but still remaining on-screen should not cause viewport-related position adjustments"?

Comment (ericmartel) [inline js/tests/unit/tooltip.js]: @cvrebert this specific test is more about "scrolling should not affect tooltip positioning".  I probably had "readjust" in mind when I typed "be adjust"

Comment (cvrebert) [inline js/tests/unit/tooltip.js]: But speaking generally, scrolling can affect tooltip position. E.g. An "auto top" tooltip will be displayed with position bottom if its associated element has been scrolled such that a top tooltip would get clipped by the top edge of the viewport.

Comment (ericmartel) [inline js/tests/unit/tooltip.js]: You are right, I guess more test coverage should be added.  This test was to catch the issue where tooltips were bound inside the browser's window even when the viewport was set as body (when normally they should have remained at the same position, next to their target)

**Comment by just-boris** [comment]
@cvrebert why it was moved to next version? It affects my app, so I'm looking forward to this fix.
Also, I can help with it, if need

**Comment by ericmartel** [comment]
@just-boris I believe the fix is working, it's code coverage in the unit tests that need to be revised.  There are many things that are not tested and I think a big rework of the whole thing would be better, but I leave it up to @cvrebert :)  Maybe you could build your own version of bootstrap including this fix in the meantime?

**Comment by just-boris** [comment]
@ericmartel of course, I tested it with my app and it helped me. But I'm not sure about other cases.

**Comment by cvrebert** [comment]
@just-boris Cases where scrolling should affect the tooltip's viewport-adjusted position. See the code review comments.

**Comment by ericmartel** [comment]
@cvrebert I will do a "manual" test tonight if I have time and eventually write unit tests for that, but the "auto" behavior should not change from this changelist, since the selection of which side to open the tooltip should show is different than repositioning it based on its content.  A lot of tests are already validating the auto placement and they all seem to pass with this PR.

**Comment by cvrebert** [comment]
@ericmartel Any luck with that additional testing?

**Comment by ericmartel** [comment]
@cvrebert hey Chris, after reading through the existing tests and validating them, it seems like there are currently 4 tests that properly validate that the scrolling of viewport elements work properly with the `auto` setting.  I think the misleading part about my PR is that my added test should have been named something along the lines of `should not adjust position when scrolling the window`, with emphasis on the window part.  That's what this PR tackles as the initial code in tooltip.js would offset the tooltip by the document's scrolling when they were on the body, instead of leaving them by their target.   The `getPosition` changes are also validated by the 4 `auto` tests.

All of this to say that, from reading the existing tests, I think the changes are fairly well covered with tests, both for explicit placement and auto placement.

If you have specific tests you would want me to add, feel free to ask!

Cheers

Eric

**Comment by cvrebert** [comment]
Okay then. Rebased again, fixed the test description, fixed conflicts related to #15933.
CC: @hnrch02 @XhmikosR for review

File: js/tooltip.js [hunk 2]
```diff
@@ -357,16 +357,34 @@
 
   }
 
+  Tooltip.prototype.getViewportBounds = function ($viewport) {
+    if ($viewport.is('body') && (/fixed|absolute/).test(this.$element.css('position'))) {
```
Comment (hnrch02) [inline js/tooltip.js]: Why is this RegExp wrapped in parens?

**Comment by hnrch02** [comment]
Otherwise LGTM.

File: js/tooltip.js (hunk 2 above)
Comment (ericmartel) [inline js/tooltip.js]: Possibly due to my C++ background, I'm still amazed with how easy it is to play with strings in js.  @cvrebert do you want me to update it?

**Comment by gharlan** [comment]
Any update on this? Can this be merged?

**Comment by Sadrak** [comment]
I tried the patch and it fixed the same problem in my project.

All other is also working like a charm.

**Comment by mdo** [comment]
Holler if there's anything you feel super strongly about here @cvrebert, but otherwise I plan on closing this along with all the other positioning related things given the v4 changes.

---

Punting as a won't fix—v4 includes a new positioning library for our affix, tooltips, and popovers.

**Pull request closed without merging**